src/Kvasir/Reconstitution/Lookups.cs

```diff
@@ -30,19 +30,30 @@ internal static partial class Lookup {
         public static object ByKey(DBData key, IEnumerable<object> entities, DataExtractionPlan extractor) {
             Debug.Assert(entities.All(e => e.GetType().IsInstanceOf(extractor.ExpectedSource)));
 
+            if (CacheDictionary.ContainsKey(key))
+                return CacheDictionary[key];
+
             foreach (var entity in entities) {
+                if (CacheValues.Contains(entity))
+                    continue;
                 var keyIdx = 0;
                 foreach (var value in extractor.ExecutePiecewise(entity)) {
                     if (!value.Equals(key[keyIdx++])) {
                         goto nextEntity;
                     }
                 }
+                CacheDictionary[key] = entity;
```

Comment:
Can you add a blank line above Line 45 here, to visually break up the logical units of the loop body?